casper/src/main/scala/coop/rchain/casper/safety/CliqueOracle.scala

```diff
@@ -0,0 +1,191 @@
+        stopper <- OptionT.liftF(
+                    getJustifications(lmAjB.message).map(
+                      js =>
+                        js.find(_.creator == lmAjB.creator)
+                          .getOrElse(lmAjB)
+                    )
+                  )
+
+        // gather all self justifications till stopper
+        potentialDisagreements <- OptionT.liftF(
+                                   DagOps
+                                     .bfTraverseF[F, M](List(lmB)) { jFromB =>
+                                       ProtoUtil
+                                         .getCreatorJustification(
+                                           jFromB,
+                                           (b: M) => b == stopper.message,
+                                           toJustification,
+                                           getJustifications
+                                         )
+                                         .map(_.map(List(_)).getOrElse(List.empty))
+                                     }
+                                     .toList
+                                 )
+
+        // if its possible that a and b disagree
+        result <- OptionT.liftF(potentialDisagreements.forallM { potentialDisagreement =>
+                   dag.isInMainChain(target, potentialDisagreement, getHeight)
+                 })
+      } yield result).fold(false)(identity)
+
+    /** across combination of validators compute pairs that do not have disagreement */
+    def computeAgreeingValidatorPairs: F[List[(V, V)]] = {
+      val pairs = agreeingWeights.keys.flatMap(v => agreeingWeights.keys.map((_, v)))
+      Stream
+        .iterable(pairs)
+        .filterNot { case (a, b) => a == b }
+        .evalFilterAsyncProcBounded {
+          case (a, b) => neverEventuallySeeDisagreement(a, b)
+        }
+        .compile
+        .toList
+    }
+
+    computeAgreeingValidatorPairs
+      .map { edges =>
+        Clique.findMaximumCliqueByWeight[V](edges, agreeingWeights)
+      }
+      .map(_.toFloat)
+  }
+
+  def normalizedFaultTolerance[F[_]: Concurrent: Log: Metrics: Span, M, V](
+      dag: DagReader[F, M],
+      target: M,
+      getWeightMap: M => F[WeightMap[V]],
+      heightF: M => F[Long],
+      latestMessageF: V => F[M],
+      toJustification: M => F[Casper.Justification[M, V]],
+      getJustifications: M => F[Set[Casper.Justification[M, V]]]
+  )(implicit show: Show[M]): F[Float] = {
+    val missingStore = Log[F]
+      .warn(s"Fault tolerance for non existing message ${target.show} requested")
+      .as(-1f)
+
+    val compute = Span[F].trace(SafetyOracleMetricsSource) {
+      for {
+        // weight map of main parent (fallbacks to message itself if no parents)
+        mainParentsWeightMap <- {
+          dag
+            .mainParent(target)
+            .flatMap {
+              case Some(mainParent) => getWeightMap(mainParent)
+              case None             => getWeightMap(target)
```

Comment:
Why stop on Option? You know you want to call `getWeightMap` on something. So resolve that first, `if not` mainParent `else` target, do getWeightMap on result.

Reply:
You mean 
```scala
dag.mainParent(target)
            .map {
              case Some(mainParent) => mainParent
              case None             => target
            }
            .map(getWeightMap)
```?